kong/plugins/ai-proxy/drivers/cohere.lua

```diff
@@ -0,0 +1,329 @@
+      }
+      this.usage = stats
+
+    elseif response_table.text then
+      -- this is a "co.chat"
+
+      this.choices[1] = {
+        index = 0,
+        message = {
+          role = "assistant",
+          content = response_table.text,
+        },
+        finish_reason = "stop",
+      }
+      this.object = "chat.completion"
+      this.model = model_info.name
+      this.id = response_table.generation_id
+  
+      local stats = {
+        completion_tokens = response_table.token_count and response_table.token_count.response_tokens or nil,
+        prompt_tokens = response_table.token_count and response_table.token_count.prompt_tokens or nil,
+        total_tokens = response_table.token_count and response_table.token_count.total_tokens or nil,
+      }
+      this.usage = stats
+  
+    else -- probably a fault
+      return nil, "'text' or 'generations' missing from cohere response body"
+  
+    end
+
+    return cjson.encode(this)
+  end,
+}
+
+function _M.from_format(response_string, model_info, route_type)
+  -- MUST return a string, to set as the response body
+  ngx.log(ngx.DEBUG, "converting from ", model_info.provider, "://", route_type, " type to kong")
+
+  if not transformers_from[route_type] then
+    return nil, fmt("no transformer available from format %s://%s", model_info.provider, route_type)
+  end
+
+  local ok, response_string, err = pcall(transformers_from[route_type], response_string, model_info)
+  if not ok or err then
+    return nil, fmt("transformation failed from type %s://%s: %s", model_info.provider, route_type, err or "unexpected_error")
+  end
+
+  return response_string, nil
+end
+
+function _M.to_format(request_table, model_info, route_type)
+  ngx.log(ngx.DEBUG, "converting from kong type to ", model_info.provider, "/", route_type)
+
+  if route_type == "preserve" then
+    -- do nothing
+    return request_table, nil, nil
+  end
+
+  local ok, response_object, content_type, err = pcall(transformers_to[route_type], request_table, model_info)
+  if not ok then
+    return nil, nil, fmt("no transformer for %s://%s", model_info.provider, route_type)
+  end
+  if err then
+    return nil, nil, fmt("error transforming to %s://%s", model_info.provider, route_type)
+  end
+
+  return response_object, content_type, nil
+end
+
+function _M.header_filter_hooks(body)
+  -- nothing to parse in header_filter phase
+end
+
+
+
+function _M.post_request(conf)
+  if ai_shared.clear_response_headers.cohere then
+    for i, v in ipairs(ai_shared.clear_response_headers.cohere) do
+      kong.response.clear_header(v)
+    end
+  end
+end
+
+function _M.pre_request(conf, body)
+  -- check for user trying to bring own model
+  if body and body.model then
+    return "cannot use own model for this instance"
+  end
+
+  return nil
+end
+
+-- returns err or nil
+function _M.configure_request(conf)
```

Comment:
I'm seeing a decent amount of code duplication in `configure_request` (and to a lesser extent, `to_format` and `from_format`. Could potentially benefit from factoring some of this out into `kong.plugins.ai-proxy.drivers.shared` or wherever is appropriate.

Reply:
I've done a little refactoring on this now, but more will come with the next round of AI Plugins, when it will also support one-shot subrequests with resty-http too.